launcher: Report actual checkup status

orelmisan opened the pull request:
Replace the hard codded report of "succeeded" and "failureReason" with an actual report.

The "succeeded" field is reported as "true" only when:
Checkup.Setup(), Checkup.Run() and Checkup.Teardown() are successfully completed.

At the moment, because the checkup results are not read, beyond the returned error from Checkup.Run(), there is not enough
data to declare the checkup as failed.

This PR also changes the launcher's tests to test the actual final report (in applicable test cases).

~~Depends on PR #45.~~

EdDev commented on kiagnose/internal/launcher/launcher.go:
```diff
@@ -52,8 +52,12 @@ func (l Launcher) Run() (runErr error) {
-		statusData.Succeeded = false
-		statusData.FailureReason = "Failure: implement me"
+		if runErr == nil {
+			statusData.Succeeded = true
```
I think this needs to be set only after the `Run()` execution.
Here, I would expect to see it set to `false` with the reason in case of an error.

In practice, there is no need to set the success boolean at all. It should be the default in case the value was not raised before.

orelmisan replied:
This code is called after the calls to `Run()` and `Teardown()`.

The success boolean is only set to `true` in case of success, and remains as the default `false` otherwise.

orelmisan replied:
Updated the commit message and the PR description to explain the logic behind "succeeded" field.

orelmisan replied:
@ormergi IMO the Status struct should stay a simple data struct.
The logic added to Launcher.Run() is a simple `if` statement.

EdDev replied:
Ok, you convinced me it is enough for now.

ormergi replied:
This defer function already bloated, it will be nicer if we can reuse some noise from it.
Note that what I suggested can also be static functions at launcher/launcher.go.

orelmisan replied:
This is only a temporary step in the way to PR #48.

ormergi replied:
Alright, fair enough.

EdDev commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -21,34 +21,103 @@ package launcher_test
+
+		checkupData := getCheckupData(t, fakeClient, configMapNamespace, configMapName)
+
+		assert.Equal(t, strconv.FormatBool(true), checkupData[reporter.SucceededKey])
+		assert.Empty(t, checkupData[reporter.FailureReasonKey])
```
Can you compare the full data? That would be preferable over multi-asserts.

orelmisan replied:
It is a bit problematic because the timestamps are set inside a struct local to the `Launcher.Run()` method.
We can expose them, but I don't think the justification is very strong.

orelmisan replied:
The actual timestamps are now zeroed before asserting the actual report equality with the expected results.

EdDev reviewed (changes requested):
Commit:launcher: Report succeeded and failureReason

I find this commit hard to review.
It includes too many changes, some not related directly with the production code addition.

Can you please perform the refactor of test structure without changing their logic (and what they test) as a pre-step? Then, it should be easy to see what was changed and what was added more easily.

As an example, I am not sure I understand what does it mean "with report" and "without report" in the failure scenarios.

EdDev reviewed (commented):
Passed through the "prefix" commits.

General note: You are using in the commit message a lot of `This` wording. I would recommend reducing its usage to the minimum. It does not add information and looks redundant.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/0a605bbaea9cfaa80f5de4cdfed48a0b81330360..6925c373534baef3367bd3412c54c9cba2efdb67): Added intermediate commits to slowly show the change to `launcher_test.go`.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/6925c373534baef3367bd3412c54c9cba2efdb67..06480039d92c2e1a1a0758cc4c0a4d0064b78f64): Changed the commit messages to reduce the usage of the word "this".

ormergi commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -40,64 +40,66 @@ func TestLauncherRunShouldSucceed(t *testing.T) {
 
-func TestLauncherShould(t *testing.T) {
-	t.Run("fail when report on checkup start is failing", func(t *testing.T) {
+func TestLauncherRunShouldFailWithReport(t *testing.T) {
```
nit: Please consider renaming to something like `TestLauncherShouldFailWhen` (and remove "when" from tests description), having functioning reporter is the default state.

orelmisan replied:
Renamed to `TestLauncherRunShouldFailWithReportWhen`.

ormergi commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -107,12 +107,14 @@ func TestLauncherRunShouldFailWithReport(t *testing.T) {
 		testLauncher := launcher.New(
 			checkupStub{},
-			&reporterStub{reportErr: errorFailOnInitialReport},
+			reporter.New(fakeClient, configMapNamespace, configMapName),
```
I think it worth to create another test function when the "production" reporter is being used, 
IMO its confusion to see one test using the "production" component and the rest uses the stub component.

orelmisan replied:
The tests division is based on whether there will be a report to the user at the end or not.

ormergi reviewed (commented):
Hi Orel, thanks for the PR.
Please see my inline comments.

orelmisan commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -178,3 +199,29 @@ func (r *reporterStub) Report(_ status.Status) error {
 	return nil
 }
+
+func checkupSpecData() map[string]string {
+	const (
+		testImageValue   = "mycheckup:v0.1.0"
+		testTimeoutValue = "1m"
+	)
+
+	return map[string]string{config.ImageKey: testImageValue, config.TimeoutKey: testTimeoutValue}
+}
```
Please see @EdDev's [comment](https://github.com/kiagnose/kiagnose/pull/45#discussion_r863616213) in PR #45.

EdDev replied:
@orelmisan , I suggested a function instead of a variable. In both cases the end result is hiding the details, and Or was talking about that.

orelmisan replied:
These details are representing the minimal checkup spec.
Why in your opinion are they important to the comprehension of the test?

ormergi replied:
It will improve visibility and make it easier to understand how the "expectedData" should look like,
The "expectedData" can be defined locally where needed.

This is nice to have..

orelmisan replied:
This could be done in a follow-up PR.

EdDev replied:
To be clear: Visibility makes sense if the data is relevant to the test. If the data is not relevant to the test, it is noise and we do not want to know about it.

EdDev commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -29,16 +29,18 @@ import (
-func TestLauncherShould(t *testing.T) {
-	t.Run("run successfully", func(t *testing.T) {
+func TestLauncherRunShouldSucceed(t *testing.T) {
+	t.Run("when checkup runs successfully", func(t *testing.T) {
```
I think the intention is to check that the launcher runs successfully, stubbing (or not) the needed to make it so.

Anyway, lets not over-complicate things... keep it simple.
There is no real need for `t.Run`, there is only one test.

And the test can be called `TestLauncherRunsSuccessfully`.

orelmisan replied:
Removed the redundant `t.Run()`.

EdDev commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -29,16 +29,18 @@ import (
 	})
+}
 
+func TestLauncherShould(t *testing.T) {
```
You can place in the name all the common wording from each test below.
I think in this case it is: `TestLauncherShouldFailWhen`.

orelmisan replied:
Changed this to:
1. `TestLauncherRunShouldFailWithReportWhen`
2. `TestLauncherRunShouldFailWithoutReportWhen`

EdDev reviewed (commented):
Reviewed commit `launcher, tests: Move success scenario to a dedicated function`.

EdDev reviewed (commented):
> launcher, tests: Split failure cases to two functions

launcher, tests: Split failure cases to two groups

> Split the failure cases into two categories:
> 1. Failures that are expected to be reported to the user.
> 2. Failures that are expected to be logged.

You need some overview so one can understand what is this all about.
Something like:
```
When the launcher execution fails, it can result in either the outcome is successfully reported to the user (through the API sever) or no (e.g. unable to access the API server).

To make this distinction clear and allow it to evolve with future scenarios, the two are grouped under different test hierarchies.
```

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/06480039d92c2e1a1a0758cc4c0a4d0064b78f64..56f078e848176071e7856385c1de1d2f043ad5e1): Rebased on PR #45.
Changed the order of the commits, so refactor commits comes first.
Kept the old test names.
Retained the reporterStub's ability to fail on the initial report.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/56f078e848176071e7856385c1de1d2f043ad5e1..917f2b820f92b6e2e523000b259b16e0c204631f): Rebase

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/917f2b820f92b6e2e523000b259b16e0c204631f..e5a5961c01da9c58c144f852223d3bce55dea342): Updated the 6th commit message and the PR description to explain the logic behind the "succeeded" field.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/e5a5961c01da9c58c144f852223d3bce55dea342..5901de3f6d8ab95b87a24849d0febf6737a09bdc): Zero the actual reported timestamps before asserting equality with expected values.

orelmisan commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -40,64 +40,66 @@ func TestLauncherRunShouldSucceed(t *testing.T) {
+}
 
-	t.Run("fail when setup and report on checkup completion are failing", func(t *testing.T) {
+func TestLauncherRunShouldFailWithoutReport(t *testing.T) {
```
Renamed this to `TestLauncherRunShouldFailWithoutReportWhen`.

EdDev commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -21,21 +21,49 @@ package launcher_test
 	assert.NoError(t, testLauncher.Run())
+
+	actualCheckupData := getCheckupData(t, fakeClient, configMapNamespace, configMapName)
+	zeroTimestamps(actualCheckupData)
```
It will be nicer if you call this on `expectedData` as well.

orelmisan replied:
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/5901de3f6d8ab95b87a24849d0febf6737a09bdc..6af425456486da401b1ce5c577bdd47f67be386e): Answered review comment.

EdDev reviewed (dismissed):
Thank you!

ormergi commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -29,16 +29,16 @@ import (
-			checkupStub{},
-			&reporterStub{},
-		)
+func TestLauncherRunsSuccessfully(t *testing.T) {
```
nit: redundant 's' at Runs.

orelmisan replied:
This is grammatically correct.

ormergi replied:
Now I realized it doesnt mention `launcher.Run()`.

ormergi commented on kiagnose/internal/launcher/launcher.go:
```diff
@@ -56,8 +56,12 @@ func (l Launcher) Run() (runErr error) {
+		} else {
+			statusData.FailureReason = runErr.Error()
+		}
+
```
regarding line [67 line](https://github.com/kiagnose/kiagnose/pull/46/commits/0516103cedea0e2314893be7b3733768688a650b#diff-57e7e8036316bd96561318e6d20d9487345e4ac53cb9c3867a324bffeb4d564eR67), it seems that that status "succeeded" is not affected by the reporter, please add a note about it in the commit message.

orelmisan replied:
The line you are referring to is the final report. 
If it fails - the user will not get a report.
There is no point in updating a value that will not be reported.

ormergi replied:
It will matter when we will log the status that is about to be written to the ConfigMap, but this is not the scope of this PR.

ormergi commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -66,32 +66,52 @@ func TestLauncherRunsSuccessfully(t *testing.T) {
+			actualCheckupData := getCheckupData(t, fakeClient, configMapNamespace, configMapName)
+			zeroTimestamps(actualCheckupData)
+
+			expectedData := checkupSpecData()
```
nit: It would be easier to understand what is the expected results if it was defined inline:
```
expectedData := map[string]string {
    reporter.SucceededKey: ..
    reporter.FailureReasonKey: ..
    config.StartTimestampKey: ..
    config.CompletionTimestampKey: ..
}
```

orelmisan replied:
I agree that it would be nicer, but `expectedData` is comprised of two maps:
1. The minimal spec.
2. All the reported data.

ormergi replied:
It seems (2) is not reflected in tests, did I miss something?

orelmisan replied:
Let me rephrase, section 2 is all the expected reported data.
Please see [here](https://github.com/kiagnose/kiagnose/pull/46/files/0516103cedea0e2314893be7b3733768688a650b..4eafc85ed21521162dceae3e2861acfd6acf589c#diff-5bfeeac3b26c7f73c22a1fa3ef5b79ca2f0f2a69eb8d9c27ce1aaf04b40af719R108-R110).

ormergi replied:
It doenst seem to be a problem, the expectedData can include what necessary in any case.

We can address it later this is nice to have.

ormergi reviewed (commented):
Regarding "launcher, tests: Check report on failure cases" commit:
What about successful tests are you going to change them to assert on the 
status that reported?

Regarding "launcher, tests: Change test when initial report fails" commit 
the commit message title is not clean, do you think you can add more info about what is going to change?

orelmisan commented:
> Regarding "launcher, tests: Check report on failure cases" commit: What about successful tests are you going to change them to assert on the status that reported?

It was done in the [previous commit](https://github.com/kiagnose/kiagnose/pull/46/commits/0516103cedea0e2314893be7b3733768688a650b) (and stated in its commit message).

> Regarding "launcher, tests: Change test when initial report fails" commit the commit message title is not clean, do you think you can add more info about what is going to change?

The commit message says:
```
Replace the reporter stub with the production reporter code,
in case when the initial report fails.
```

What are you missing?

EdDev commented:
> > Regarding "launcher, tests: Change test when initial report fails" commit the commit message title is not clean, do you think you can add more info about what is going to change?
> 
> The commit message says:
> 
> ```
> Replace the reporter stub with the production reporter code,
> in case when the initial report fails.
> ```
> 
> What are you missing?

Maybe he is missing the reasoning behind it.
Also, I am not sure if `in case when the initial report fails` is a valid sentence (that `when` part does not fit).

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/6af425456486da401b1ce5c577bdd47f67be386e..50872a90de1cc4f10dff18d41f3f40b85446b193): Updated last commit message.

EdDev commented:
> This is done in order to use the production reporter whenever possible.

I am not sure if this qualifies to reasoning.
How about this commit:
```
launcher, tests: Use production reporter instead of stub

The test that covers the scenario in which the report fails to
post the start timestamp is currently using a reporter stub.

In order to cover a more realistic scenario (and setup), the
stub is replaced with the production reporter.
```

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/50872a90de1cc4f10dff18d41f3f40b85446b193..6f563a7ba21053ef518ffa931586f6f0001831c4): Changed last commit message.

ormergi reviewed (approved):
Thanks
